LetsDoStuff.WebApi/Services/SubscribingManager.cs

```diff
@@ -0,0 +1,106 @@
+{
+    public class SubscribingManager : ISubscribingService
+    {
+        private readonly LdsContext db;
+
+        public SubscribingManager(LdsContext context)
+        {
+            db = context;
+        }
+
+        public void MakeUserSubscribedToActivityByEmailAndId(string emailUser, int idActivity)
+        {
+            User subscriber = FindSubscriberByEmail(emailUser);
+
+            Activity activityForSubscribing = GetActivityOrNull(idActivity)
+                                                        ?? throw new ArgumentException($"Activity with id {idActivity} has not been found");
+
+            if (IsThereActivityinUserPartis(subscriber, activityForSubscribing))
+            {
+                throw new ArgumentException($"{subscriber.Email} has already subscribed to the event with id {idActivity}");
+            }
+
+            subscriber.Partis.Add(activityForSubscribing);
+
+            db.SaveChanges();
+        }
+
+        public void MakeUserUnsubscribedToActivityByEmailAndId(string emailUser, int idActivity)
+        {
+            User subscriber = FindSubscriberByEmail(emailUser);
+
+            Activity activityForUnsubscribing = GetActivityOrNull(idActivity)
+                                        ?? throw new ArgumentException($"Activity with id {idActivity} has not been found");
+
+            if (!IsThereActivityinUserPartis(subscriber, activityForUnsubscribing))
+            {
+                throw new ArgumentException($"{subscriber.Email} hasn't had subscription to the event with id {idActivity}");
+            }
+
+            subscriber.Partis.Remove(activityForUnsubscribing);
+            db.SaveChanges();
+            return;    
+        }
+
+        public List<ActivityResponse> GetAllActivitiesOfTheUser(string userName)
+        {
+            var response = db.Users.AsNoTracking()
+               .Where(u => u.Email == userName)
+               .Include(u => u.Partis)
+               .ThenInclude(a => a.Creator)
+               .Include(u => u.Partis)
+               .ThenInclude(a => a.Tags)
+               .FirstOrDefault()
+               .Partis
+               .Select(a => new ActivityResponse()
+               {
+                   Id = a.Id,
+                   Name = a.Name,
+                   Description = a.Description,
+                   Capacity = a.Capacity,
+                   Creator = new ActivityCreatorResponse()
+                   {
+                       Id = a.Creator.Id,
+                       Name = a.Creator.FirstName,
+                       Login = a.Creator.Email
+                   },
+                   Tags = a.Tags.Select(t => t.Name).ToList()
+               });
+
+            return response.ToList();
+        }
+
+        #region
+        private User FindSubscriberByEmail(string emailUser) => db.Users.Where(h => h.Email == emailUser)
+                                   .Include(u => u.Partis)
+                                   .Single();
```

Comment:
Here it's not necessary but It's more suited for introducing attention: we use email here only because it unique ( and because of authorization of cause )

Reply:
yes, it is. but it is twice slower on average and using it makes sense only if the check of singularity has to be done. In general, the best method for such cases is `FirstOrDefault()`

Reply:
> yes, it is. but it is twice slower on average and using it makes sense only if the check of singularity has to be done. In general, the best method for such cases is `FirstOrDefault()`

I could agree that in general FirstOrDefault is more useful but using FirstOrDefault() we are always obliged to check a returning parameter on the default and be aware of what a default means ( could be null or just an empty line or etc). So if I don't want to check this and prefer to throw an average Exception ( and if a DB isn't enormously big ) I prefer to use Single or First for Fail-fast.

Reply:
In LinqToEF the default can only be null, so the best way to handle it is a good human-readable exception message like
`db.Users.Where(h => h.Email == emailUser)
                                   .Include(u => u.Partis)
                                   .FirstOrDefault()
                                   ?? throw new ArgumentException($"There is no {emailUser}");`

Reply:
> In LinqToEF the default can only be null, so the best way to handle it is a good human-readable exception message like
> `db.Users.Where(h => h.Email == emailUser) .Include(u => u.Partis) .FirstOrDefault() ?? throw new ArgumentException($"There is no {emailUser}");`

But here's the thing, I'm using throw new ArgumentException bellow for informing a user about his mistakes... not mine ( or other programmer's mistakes ) or about an exceptional situation. But this is not an exception, this is someone's mistake ( because why a user could sign in successfully in the first place?! ), and I want to underline this fact at that place that the mistake turned up ( not at a controller by try/catch structure ). But ok, if you insist on this I'll replace Single() on FirstOrDefault(). Just let me know by thumbing up.